### DISCUSSION TITLE
Can someone explain what are the meanings of these timings

### DISCUSSION THREAD

**Discussion by Tommy787576** [ggml-org/llama.cpp#1323] [General]
![image](https://user-images.githubusercontent.com/70518894/236243926-f85e2a8e-8fb8-443b-9106-98a6c60c2490.png)
What does load / sample / prompt eval / eval / total time mean? What does 0.58 ms per **run** mean? Why can't I see the timings report like what readme shows?
![image](https://user-images.githubusercontent.com/70518894/236244304-2b48d19d-1691-42af-92e6-37bcfa554912.png)
This is much clearer! What is the mapping between them? I want to know the predict: token per ms.

**Comment by akx** [comment]
The timing print format was changed in #370, if that helps.

**Reply by Tommy787576** [reply]
Thank you

**Reply by IdoAmit198** [reply]
I tried to follow the Change mentioned above and couldn't find the answer there..
If any of you can explain how to understand the time per token from the new print format, I'd really appreciate it.

**Comment by SlyEcho** [comment]
```
llama_print_timings:        load time =   134.84 ms
llama_print_timings:      sample time =    12.25 ms /    32 runs   (    0.38 ms per token)
llama_print_timings: prompt eval time =   925.78 ms /    24 tokens (   38.57 ms per token)
llama_print_timings:        eval time =  4178.32 ms /    31 runs   (  134.78 ms per token)
llama_print_timings:       total time =  5211.06 ms
```

- load time: loading model file
- sample time: ~~generating tokens from the prompt/file~~ choosing the next likely token.
- prompt eval time: how long it took to process the prompt/file by LLaMa before generating new text.
- eval time: how long it took to generate the output (until `[end of text]` or the user set limit).
- total: all together

When it says 38.57 ms per token, it means it evaluated 24 tokens in 925.78 ms. Lower is better.

Another common metric is tokens per second which is used in other ML projects than llama.cpp. You can get that by calculating 1000/38.57 = 25.93 t/s. Higher is better.

Why is eval time slower than prompt eval time?

Because we can only predict the single next token, that means that after the sampler chooses the next token the model has to be run again but with a batch size of 1. In the prompt eval phase, the model can evaluate large batches (512 max) meaning less overhead and more efficiency, especially with BLAS or GPU.

EDIT: "sample time" is not tokenization time, it is actually the "sampling" time, that means running the RNG, sorting and filtering candidates, etc.

**Reply by IdoAmit198** [reply]
@SlyEcho Thanks a lot, it is very informative :)

**Reply by isaac-chung** [reply]
Summing all timings together gave 5251.19ms instead of the printed 5211.06ms. I consistently see this in my trials as well, where the actual summed time > printed total time. 
Any idea where the difference is coming from?

**Reply by JayKayNJIT** [reply]
same here, I am manually summing all the times and they never matches bottom line, any idea??

**Reply by joyyang1215** [reply]
Many thanks

**Reply by yangxile** [reply]
I came across a different explaination of "sample time" in https://github.com/ggerganov/llama.cpp/issues/2237, which says "sample time: time it takes to "tokenize" (sample) the prompt message for it to be processed by the program.", which is very confusing, can anyone help elaborate on this "EDIT: "sample time" is not tokenization time, it is actually the "sampling" time, that means running the RNG, sorting and filtering candidates, etc.", it is the time to decide the next token? or just system initialization time?

**Reply by adb1997** [reply]
Did anyone figure out why the value of total time differs from manually adding them individually?
